statistics/databuffer.go

```diff
@@ -0,0 +1,379 @@
+/*
+http://www.apache.org/licenses/LICENSE-2.0.txt
+
+Copyright 2016 Intel Corporation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package statistics
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/intelsdi-x/snap-plugin-lib-go/v1/plugin"
+)
+
+//dataBuffer contains 2 slices that point to buffer values:
+type dataBuffer struct {
+	data                      buffer //data is sorted by value
+	dataByTimestamp           buffer //dataByTimestamp is sorted by timestamp
+	index, slidingFactorIndex int    //sliding factor specifies how many data values to include over each sliding window
+}
+type buffer []*data
+
+//data holds the timestamp and the value (actual data)
+type data struct {
+	ts    time.Time //timestamp
+	value float64   //data
+}
+
+type Flag struct {
+	Count, Sum, Mean, Median, Minimum, Maximum, Range, Variance,
+	StandardDeviation, Mode, Kurtosis, Skewness, Trimean,
+	FirstQuartile, ThirdQuartile, QuartileRange,
+	SecondPercentile, NinthPercentile, TwentyFifthPercentile, SeventyFifthPercentile,
+	NinetyFirstPercentile, NinetyFifthPercentile, NinetyEighthPercentile,
+	NinetyNinthPercentile bool
+}
+
+//functions to sort data
+func (a buffer) Len() int           { return len(a) }
+func (a buffer) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a buffer) Less(i, j int) bool { return a[i].value < a[j].value }
+
+//Inserts data in a sorted order
+func (b *dataBuffer) Insert(value float64, ts time.Time) {
+	if len(b.data) < cap(b.data) {
+		element := &data{value: value, ts: ts}
+		b.data = append(b.data, element)
+		b.dataByTimestamp = append(b.dataByTimestamp, element)
+	} else {
+		b.dataByTimestamp[b.index].ts, b.dataByTimestamp[b.index].value = ts, value
+		b.index = (b.index + 1) % cap(b.data)
+	}
```

Comment:
I'm not sure what is your thought, so here is a sum up of what's happening.

`b.data` is sorted by value.
`b.dataByTimestamp` is sorted by timestamp.

All element in `b.data` are linked to element in `b.dataByTimestamp`, because those are slice of pointer.

When we sort `b.data` , only one element is not sorted. Sorting is actually super fast.
Having a circular array for `b.dataByTimestamp` avoid relocating memory when inserting an element.

We keep timestamps in order to know what sliding windows is returned (through tags).